### PR TITLE
Remove extra graph breaks

### LINKED ISSUE #104053: Tracking issue for optimizer graph not being an inference graph
### 🐛 Describe the bug

The optimizer is incorrectly classified as a training graph in aot. The main reason this occurs is that some parameters have `requires_grad=True` *and* torchdynamo calls set_grad_enabled(True|False) while simultaneously putting a call_function node in the graph which performs the global state change. 

As a result, when dynamo traces a no_grad context manager, it will trace the *exit* of the context manager *before* calling the user compiler. As a result, by the time we reach aot, grad is enabled, and the graph is treated as a training graph.

Why this matters:
Functionalization appends an epilogue which aggregates all mutation updates of tensors into a single code block consisting of a bunch of copies. In inference mode, these copies are visible to inductor while in training mode, these copies are not visible to inductor. If the copies are not visible to inductor, inductor will allocate tensors for every single parameter in the graph in order for them to get sent to the epilogue, which is atrocious for perf. 

Solutions:
1. graph break before grad is re-enabled
2. in functionalization make the epilogue visible in training mode
3. with compiled backwards, the entire graph is an inference graph by default

1 is two lines of code, so that approach is the easiest. 
3 is getting implemented currently, so once it's ready we can remove the graph breaks.

### Error logs

N/A

### Minified repro

_No response_

### Versions

N/a

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @zou3519

### PR DESCRIPTION
**Pull request by mlazos**
Fixes https://github.com/pytorch/pytorch/issues/104053

cc @albanD @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @janeyx99 @aakhundov

### CODE REVIEW

**Review by janeyx99 — COMMENTED** [review]
Wooo!!! Happy to see this change!

How do we know this doesn’t make tracing slow (an original concern)? 

And I’d anticipate 2 graph breaks fewer from the nested functions, but it seems like they’re mostly 1?

And trusting u to remove all references of the commenting from the codebase.

**Comment by mlazos** [comment]
> Wooo!!! Happy to see this change!
> 
> How do we know this doesn’t make tracing slow (an original concern)?
> 
> And I’d anticipate 2 graph breaks fewer from the nested functions, but it seems like they’re mostly 1?
> 
> And trusting u to remove all references of the commenting from the codebase.

The original reason for these graph breaks is not tracing speed - it was no_grad detection being inadequate in AOT but it looks good now (see linked issue). No other references since this graph break was at the one decorator. It's one graph break I think because there's another graph break immediately after this one with no code in between.

**Review by janeyx99 — APPROVED** [review]

**Comment by sanchitintel** [comment]
Hi, perhaps a silly question, but can the graph-break corresponding to `optimizer.step` be avoided by having Inductor codegen a kernel that could handle `optimizer.step`'s compute? Thanks!

**Comment by mlazos** [comment]
> Hi, perhaps a silly question, but can the graph-break corresponding to `optimizer.step` be avoided by having Inductor codegen a kernel that could handle `optimizer.step`'s compute? Thanks!

We handle the compute already, this graph break is for a more complex issue where there are extra copies due to the graph being a training graph as opposed to an inference graph. This hasn't been landed because when these graph breaks are removed there's still a catastrophic perf degradation.
